### DISCUSSION TITLE
Marker label based on data type

### DISCUSSION THREAD

**Discussion by ivaylopenevmg** [patrykandpatrick/vico#974] [Questions]
### Question

Hello, in my implementation I need to show a label that represents the data type in the marker along side with the numeric value.

![Screenshot 2024-12-16 at 17 23 06](https://github.com/user-attachments/assets/cb029f5e-3ae1-4a85-b0f5-01bbb9fe41f0)

Looking at the source code the object we can work with is:

```
public data class Point(
    val entry: LineCartesianLayerModel.Entry,
    val canvasY: Float,
    val color: Int,
  )
```

So there is no way to map a label to a `Point`, the work around I use is to map labels with the colors. Example:

```
fun getDataNameFromColor(color: Int): String {
    return when (color) {
        -8832, -672691 -> "Grid"
        -8735836 -> "Solar"
        -12488092 -> "Injection"
        ...
        else -> ""
    }
}
```

Is there a better way to do this?

### Vico version(s)

2.0.0-beta.1

### UI framework(s)

Jetpack Compose

**Comment by Gowsky** [comment]
Hello, @ivaylopenevmg. There are more robust ways to do this, but the best one depends on the specifics.

Do the series have the same _x_ values, or is it possible for one series to have a gap where another one doesn't have it, leading to a case where e.g. there are three lines, but the marker highlights only two points at a given _x_ value?

**Reply by ivaylopenevmg** [reply]
Hello, @Gowsky we have no gaps in the series, and they can have same x values. Zero values can also be possible.

**Reply by Gowsky** [reply]
In that case, you can do this based on the order of the elements in `targets`. Targets are guaranteed to be added in the order of the series. Extras may be needed if series are shown/hidden dynamically.

Adding a `seriesIndex` property to the `Entry` classes was brought up previously. It's not needed here, but it could make things slightly simpler. Feel free to open a feature request if you'd find this helpful.

**Reply by ivaylopenevmg** [reply]
Hello @Gowsky thank you for your reply. We do show/hide series dynamically. Could you please clarify on what do you mean by "Extras"?

**Reply by Gowsky** [reply]
Extras are covered [in the guide](https://www.patrykandpatrick.com/vico/guide/stable/android/core/cartesian-charts/cartesian-chart-model-producer#extras). Let us know if you have any specific questions.

**Reply by Inky-developer** [reply]
Hello @Gowsky,

> Do the series have the same x values, or is it possible for one series to have a gap where another one doesn't have it, leading to a case where e.g. there are three lines, but the marker highlights only two points at a given x value?

Unfortunately I have this situation, where it is not guaranteed that there are the same number of points per x-coordinate. How would the extra-based approach to fix this work? Unfortunately the guide returns a 404

**Reply by Gowsky** [reply]
Hello, @Inky-developer. Here's the updated [link](https://www.patrykandpatrick.com/vico/guide/stable/android/core/cartesian-charts/cartesian-chart-model-producer#extra).

**Reply by Inky-developer** [reply]
Thanks for the updated link. Unfortunately I still don't fully understand: How can this be used to associate data with the `Target`s?

**Reply by Gowsky** [reply]
As per my previous comment, targets are guaranteed to be added to `targets` in the order of the series. This means that given a list of targets for a given _x_-value, you know that the first one corresponds to the first series with an entry at this _x_-value, the second one corresponds to the second series with an entry at this _x_-value, and so on. You need a simple extra that can tell you what these series are.

**Reply by Inky-developer** [reply]
> you know that the first one corresponds to the first series with an entry at this x-value

I don't think this is necessarily true, since a point is not guaranteed to contain values for every series. For example, if I have two series but one series does not contain a value for the given marker position, the the `targets` list will only contain one value, and AFAIK there is no way to figure out to which series this entry corresponds.

**Reply by Gowsky** [reply]
> For example, if I have two series but one series does not contain a value for the given marker position, the the `targets` list will only contain one value

Yes, and whether that is the first series or the second series from the model, it is the first series **with an entry at this _x_-value**, as per my comment.

> AFAIK there is no way to figure out to which series this entry corresponds

As mentioned above, you can do it using extras. When adding the series, you know for which _x_-values they have entries. You need to save this information as an extra (e.g., as a simple map). Then, continuing your example, if `targets` has one element, you look at its _x_ value, check the extra, and determine which series has an entry there. That’s the highlighted series.

**Reply by Inky-developer** [reply]
Ah okay, thanks for explaining

**Comment by ivaylopenevmg** [comment]
@Gowsky we are still facing issues with this. Any chance we can schedule a 10min call?

**Reply by Gowsky** [reply]
Could you describe what the issue is?

**Reply by ivaylopenevmg** [reply]
@Gowsky the issues is the one described on: "[Marker label based on data type #974](https://github.com/patrykandpatrick/vico/discussions/974#)". We need to display data labels in the marker and the current SDK doesn't have a solution for that. We tried the "Extras" approach but it still not what we need. We feel like the best way to show you the problem would be a very quick call where you might give us a solution or simply reach to a conclusion that there is a limitation. Would you be able to give us 10 min of your time this week?

**Reply by Gowsky** [reply]
> the issues is the one described on: "Marker label based on data type #974".

The problem statement from the original post ("There is no way to map a label to a `Point`") is inaccurate — as per my comments, it **is** possible to achieve what you need via indices and extras. If, with this knowledge, you're still facing issues, then it must be something else — related to indices or extras. I understand the context and the desired result. I'm asking about the **exact** problem at this stage, since the original question has already been answered, and a general solution has been provided.

> and the current SDK doesn't have a solution for that

As described above, it does. It's not helpful to counter a description of the solution by saying that no solution exists. The goal you've described **can** be achieved using the means I've mentioned. If you need any clarification on my suggestions, feel free to ask.

> We tried the "Extras" approach but it still not what we need.

Based on the description of the desired result, extras are exactly what's needed here, in combination with indices. If you're having trouble implementing the behavior you described using extras, please show us what you've tried so that we can help. This is one example of a description of the exact problem.

> We feel like the best way to show you the problem would be a very quick call where you might give us a solution or simply reach to a conclusion that there is a limitation. Would you be able to give us 10 min of your time this week?

This is a free and open-source project, so we're generally unable to provide such live consultancy. An exception may be made, but we'd still need more concrete information first.
